x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/aggs/heuristic/PValueScoreTests.java

```diff
@@ -0,0 +1,146 @@
+        boolean backgroundIsSuperset = randomBoolean();
+        long supersetCount = randomLongBetween(0L, Long.MAX_VALUE/2);
+        long subsetCount = randomLongBetween(0L, supersetCount);
+        if (backgroundIsSuperset) {
+            supersetCount += subsetCount;
+        }
+
+        PValueScore pValueScore = new PValueScore(backgroundIsSuperset);
+        assertThat(pValueScore.getScore(subsetCount, subsetCount, subsetCount, supersetCount), greaterThanOrEqualTo(700.0));
+    }
+
+    public void testLowPValueScore() {
+        boolean backgroundIsSuperset = randomBoolean();
+        long supersetCount = randomLongBetween(0L, Long.MAX_VALUE/2);
+        long subsetCount = randomLongBetween(0L, supersetCount);
+        long subsetFreqCount = randomLongBetween(0L, subsetCount/5);
+        if (backgroundIsSuperset) {
+            supersetCount += subsetCount;
+        }
+
+        PValueScore pValueScore = new PValueScore(backgroundIsSuperset);
+        assertThat(
+            pValueScore.getScore(subsetFreqCount, subsetCount, subsetCount, supersetCount),
+            allOf(lessThanOrEqualTo(5.0), greaterThanOrEqualTo(0.0))
+        );
+    }
+
+    public void testPValueScore() {
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 100, 100, 1000)),
+            closeTo(1.0, eps)
+        );
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 100, 10, 1000)),
+            closeTo(0.0014942974424670364, eps)
+        );
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 100, 200, 1000)),
+            closeTo(1.0, eps)
+        );
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(20, 10000, 5, 10000)),
+            closeTo(0.3154715149153073, eps)
+        );
+    }
+
+    public void testSmallChanges() {
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(1, 4205, 0, 821496)),
+            closeTo(0.47862401010222105, eps)
+        );
+        // Same(ish) ratios
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 4205, 195, 82149)),
+            closeTo(0.4946943227464169, eps)
+        );
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(10, 4205, 1950, 821496)),
+            closeTo(0.49338445847730966, eps)
+        );
+
+        // 4% vs 0%
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(168, 4205, 0, 821496)),
+            closeTo(6.340459324365642e-27, eps)
+        );
+        // 4% vs 2%
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(168, 4205, 16429, 821496)),
+            closeTo(2.392323732118125e-06, eps)
+        );
+        // 4% vs 3.5%
+        assertThat(
+            FastMath.exp(-new PValueScore(false).getScore(168, 4205, 28752, 821496)),
+            closeTo(0.2364469224974871, eps)
+        );
+    }
+
```

Comment:
How about adding some invariant tests on random input that:
1. The score is always >= 0,
2. If the fraction on foreground set is less than fraction on background set the score is identically 0,
3. Increasing subsetFreq for all other parameters fixed increases score,
4. Increasing supersetFreq for all other parameters fixed decreases score.

It might also be nice to check it does actually throw when values will overflow.

Reply:
There are super class tests on score consistency that handle:

> 1. The score is always >= 0,
> 4. Increasing supersetFreq for all other parameters fixed decreases score.

I will add some for 2, 3